lib/extras/gain_map.cc

```diff
@@ -0,0 +1,216 @@
+  };
+};
+
+template <size_t N>
+void* FixedSizeMemoryManagerAlloc(void* opaque, size_t capacity) {
+  auto manager = static_cast<FixedSizeMemoryManager<N>*>(opaque);
+  if (capacity > N + jxl::memory_manager_internal::kAlias) {
+    return nullptr;
+  }
+  return manager->memory_;
+}
+
+}  // namespace
+
+JXL_BOOL JxlGainMapGetBundleSize(const JxlGainMapBundle* map_bundle,
+                                 size_t* bundle_size) {
+  if (map_bundle == nullptr) return 0;
+
+  FixedSizeMemoryManager<sizeof(jxl::ColorEncoding)> memory_manager;
+  jxl::ColorEncoding internal_color_encoding;
+  jxl::BitWriter color_encoding_writer(memory_manager.memory_manager());
+  if (map_bundle->has_color_encoding) {
+    JXL_RETURN_IF_ERROR(
+        internal_color_encoding.FromExternal(map_bundle->color_encoding));
+    if (!jxl::Bundle::Write(internal_color_encoding, &color_encoding_writer,
+                            /*layer=*/0, nullptr)) {
+      return JXL_FALSE;
+    }
+  }
+  color_encoding_writer.ZeroPadToByte();
+
+  *bundle_size =
+      1 +                                   // size of jhgm_version
+      2 +                                   // size_of gain_map_metadata_size
+      map_bundle->gain_map_metadata_size +  // size of gain_map_metadata
+      1 +                                   // size of color_encoding_size
+      color_encoding_writer.GetSpan().size() +  // size of the color_encoding
+      4 +                                       // size of compressed_icc_size
+      map_bundle->alt_icc_size +                // size of compressed_icc
+      map_bundle->gain_map_size;                // size of gain map
+  return JXL_TRUE;
+}
+
+JXL_BOOL JxlGainMapWriteBundle(const JxlGainMapBundle* map_bundle,
+                               uint8_t* output_buffer,
+                               size_t output_buffer_size,
+                               size_t* bytes_written) {
+  if (map_bundle == nullptr) return 0;
+
+  uint8_t jhgm_version = map_bundle->jhgm_version;
+
+  FixedSizeMemoryManager<sizeof(jxl::ColorEncoding)> memory_manager;
+  jxl::ColorEncoding internal_color_encoding;
+  jxl::BitWriter color_encoding_writer(memory_manager.memory_manager());
+  if (map_bundle->has_color_encoding) {
+    JXL_RETURN_IF_ERROR(
+        internal_color_encoding.FromExternal(map_bundle->color_encoding));
+    if (!jxl::Bundle::Write(internal_color_encoding, &color_encoding_writer,
+                            /*layer=*/0, nullptr)) {
+      return JXL_FALSE;
+    }
+  }
+
+  color_encoding_writer.ZeroPadToByte();
+
+  size_t cursor = 0;
+  if (cursor + 1 <= output_buffer_size) {
+    memcpy(output_buffer + cursor, &jhgm_version, 1);
+    cursor += 1;
+  }
+
+  uint16_t metadata_size_le = JXL_BSWAP16(map_bundle->gain_map_metadata_size);
+  if (cursor + 2 <= output_buffer_size) {
+    memcpy(output_buffer + cursor, &metadata_size_le, 2);
+    cursor += 2;
+  }
+
+  if (cursor + map_bundle->gain_map_metadata_size <= output_buffer_size) {
```

Comment:
Not fundamental in the encoder, but ideally those operations should check for overflow.

Reply:
I added the use of `SafeAdd`, which we have in common.h. 
Is this what you had in mind?

Reply:
now also actually returning an error